### DISCUSSION TITLE
Using block json

### DISCUSSION THREAD

**Discussion by Hesesses** [nathanheffley/laravel-slack-blocks#16] [General]
Hello and thank you for creating this.

I was wondering why we cant just use the slack json as attachment instead of using ->type(...) ->text(...) etc

**Comment by nathanheffley** [comment]
Hey there! I moved this to discussions since it's more of a general question.

The reason I have all the helper functions is because writing all the JSON and formatting it properly can be tricky. The helper functions provide type-hinting so that it prompts the developer letting them know what they can add to a block, doing all the formatting for them, etc.

That being said, I think the idea of providing a way to just use raw JSON would be good. It would allow copying the JSON from Slack's nifty block builder tool, and if there's a new feature added to the API, a savvy user could write the JSON for it before it makes it into this package.

Could you write some pseudo-code in the way you would like it to work and paste it here?

**Comment by webdevnerdstuff** [comment]
I would also love it would have this ability, as the Block Kit Builder uses JSON. Due to that we cannot use it. A good usage would be something like the following:

```
$data = [
    "blocks" => [
        [
            "type" => "section",
            "text" => [
                "type" => "mrkdwn",
                "text" => "Hello World",
            ]
        ]
    ]
];

(new SlackMessage)->blockKitBuilder(json_encode($data));
```

There is another package that does something like this, but I prefer not changing what I'm already using. I won't post the link as I think it's usually in bad taste to post links to other competing packages. Unless you are ok with it.

**Reply by nathanheffley** [reply]
Hey @webdevnerdstuff! I am a bit confused by your comment, do you mean that you use the Slack block kit builder tool that Slack provides to generate your JSON and would prefer to not translate that into the functions that this package provides?

As I said earlier, I am more than happy to merge in a generic `json` function that accepts any string you pass to it, so that more advanced users can use raw JSON (or JSON generated by the builder tool).

I don't know when I'd have time to write the PR myself, so if you'd like this feature I would really appreciate if you take it upon yourself. The way I would like to see it implemented would be to add a `json` property to https://github.com/nathanheffley/laravel-slack-blocks/blob/a0b83652c7da955445284a47e603201c2f83252d/src/Messages/SlackMessage.php and then in https://github.com/nathanheffley/laravel-slack-blocks/blob/2.0/src/Channels/SlackWebhookChannel.php#L30 check to see if the message has a JSON payload, and if it does just return that payload instead of building the payload from the blocks and attachments.

As long as the PR is close to that I could clean it up and get it merged ASAP.

**Reply by webdevnerdstuff** [reply]
You are not confused, as it sounds like you are explaining exactly what I meant. :) 

When I have some free time I can maybe take a look at taking a stab at getting it working. This will be more uncharted area for me as I haven't written a package like this one before. I've done more JavaScript npm packages/plugins.

**Reply by nathanheffley** [reply]
It should be pretty straightforward hopefully, here is an article that shows how (after cloning this repo to be in a folder beside your project) you can include a local directory as the package instead of pulling the real package from Packagist. That way you can edit the package code locally until it works how you expect: https://jackwhiting.co.uk/posts/working-with-local-packages-in-laravel/

Let me know if you run into any issues developing on the package locally.
